### PR TITLE
IGNITE-6167: Ability to enabled TLS protocols and cipher suites

### PR DESCRIPTION
**Pull request by mcherkasov**

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/ssl/SslContextFactory.java [hunk 1]
```diff
@@ -89,6 +90,12 @@
     /** Trust managers. */
     private TrustManager[] trustMgrs;
 
+    /** Enabled cipher suites. */
+    private String[] cipherSuites;
+
+    /** Enabled cipher suites. */
```
Comment (alamar) [inline modules/core/src/main/java/org/apache/ignite/ssl/SslContextFactory.java]: We already have sslContextFactory in igniteConfiguration which has setProtocol().
Having protocols here is redundant and it is not obvious which one of two will be used. I suggest dropping it, using new String[] {sslContextFactory.create().getProtocol()} in its place

File: modules/core/src/main/java/org/apache/ignite/ssl/SslContextFactory.java [hunk 2]
```diff
@@ -310,6 +349,18 @@ private SSLContext createSslContext() throws SSLException {
 
             SSLContext ctx = SSLContext.getInstance(proto);
 
+            if(cipherSuites != null || protocols != null) {
```
Comment (alamar) [inline modules/core/src/main/java/org/apache/ignite/ssl/SslContextFactory.java]: A whitespace is desired

File: modules/core/src/main/java/org/apache/ignite/ssl/SslContextFactory.java [hunk 3]
```diff
@@ -310,6 +349,18 @@ private SSLContext createSslContext() throws SSLException {
...
+                SSLParameters sslParameters = new SSLParameters();
+
+                if(cipherSuites != null)
```
Comment (alamar) [inline modules/core/src/main/java/org/apache/ignite/ssl/SslContextFactory.java]: A whitespace is desired

File: modules/core/src/main/java/org/apache/ignite/ssl/SslContextFactory.java [hunk 4]
```diff
@@ -310,6 +349,18 @@ private SSLContext createSslContext() throws SSLException {
...
+                    sslParameters.setCipherSuites(cipherSuites);
+
+                if(protocols != null)
```
Comment (alamar) [inline modules/core/src/main/java/org/apache/ignite/ssl/SslContextFactory.java]: A whitespace is desired

File: modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySslParametersTest.java [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+    private volatile String[] cipherSuites;
+
+    /** */
+    private volatile String[] protocols;
+
+    /** {@inheritDoc} */
+    @Override protected IgniteConfiguration getConfiguration(String gridName) throws Exception {
+        IgniteConfiguration cfg = super.getConfiguration(gridName);
+
+        SslContextFactory factory = (SslContextFactory)GridTestUtils.sslTrustedFactory("node01", "trustone");
```
Comment (alamar) [inline modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySslParametersTest.java]: Split into two lines is desired

File: modules/core/src/main/java/org/apache/ignite/ssl/SSLSocketFactoryWrapper.java [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+
+    /** */
+    private final SSLSocketFactory delegate;
+
+    /** */
+    private final SSLParameters parameters;
+
+    /** */
+    SSLSocketFactoryWrapper(SSLSocketFactory delegate,
+        SSLParameters parameters) {
```
Comment (alamar) [inline modules/core/src/main/java/org/apache/ignite/ssl/SSLSocketFactoryWrapper.java]: Extra line feed

File: modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySslParametersTest.java [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+import org.apache.ignite.IgniteCheckedException;
+import org.apache.ignite.configuration.IgniteConfiguration;
+import org.apache.ignite.ssl.SslContextFactory;
+import org.apache.ignite.testframework.GridTestUtils;
+import org.apache.ignite.testframework.junits.common.GridCommonAbstractTest;
+
+/**
+ * Tests cases when node connects to cluster with different set of cipher suites.
+ */
+public class TcpDiscoverySslParametersTest extends GridCommonAbstractTest {
```
Comment (alamar) [inline modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySslParametersTest.java]: Would be nice if nonexistent ciphers and protocols were tested also.

File: modules/core/src/main/java/org/apache/ignite/ssl/SslContextFactory.java (hunk 1 above)
Comment (mcherkasov) [reply modules/core/src/main/java/org/apache/ignite/ssl/SslContextFactory.java]: it's not the same, the protocol field allows to set only one particular
protocol, while protocols allow you to set several different versions, for
example with protocols you can set:
TLSv1.1 and TLSv1.2, but exclude TLSv1.0

File: modules/core/src/main/java/org/apache/ignite/ssl/SslContextFactory.java (hunk 2 above)
Comment (mcherkasov) [reply modules/core/src/main/java/org/apache/ignite/ssl/SslContextFactory.java]: fixed

File: modules/core/src/main/java/org/apache/ignite/ssl/SslContextFactory.java (hunk 3 above)
Comment (mcherkasov) [reply modules/core/src/main/java/org/apache/ignite/ssl/SslContextFactory.java]: fixed

File: modules/core/src/main/java/org/apache/ignite/ssl/SslContextFactory.java (hunk 4 above)
Comment (mcherkasov) [reply modules/core/src/main/java/org/apache/ignite/ssl/SslContextFactory.java]: fixed

File: modules/core/src/main/java/org/apache/ignite/ssl/SSLSocketFactoryWrapper.java (hunk 6 above)
Comment (mcherkasov) [reply modules/core/src/main/java/org/apache/ignite/ssl/SSLSocketFactoryWrapper.java]: fixed

File: modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySslParametersTest.java (hunk 7 above)
Comment (mcherkasov) [reply modules/core/src/test/java/org/apache/ignite/spi/discovery/tcp/TcpDiscoverySslParametersTest.java]: added

**Pull request closed without merging**